slurmutils/models/model.py

```diff
@@ -0,0 +1,116 @@
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Base classes and methods for composing Slurm data models."""
+
+__all__ = ["BaseModel", "LineInterface", "format_key", "generate_descriptors"]
+
+import copy
+import json
+import re
+from abc import ABC, abstractmethod
+from typing import Any, Dict
+
+from ..exceptions import ModelError
+
+_acronym = re.compile(r"(?<=[A-Z])(?=[A-Z][a-z])")
+_camelize = re.compile(r"(?<=[a-z0-9])(?=[A-Z])")
+
+
+def format_key(key: str) -> str:
+    """Format Slurm configuration keys from SlurmCASe to camelCase.
+
+    Args:
+        key: Configuration key to format into camel case.
+
+    Notes:
+       Slurm configuration syntax does not follow proper PascalCasing
+       format, so we cannot put keys directly through a kebab case converter
+       to get the desired format. Some additional processing is needed for
+       certain keys before the key can properly camelized.
+
+       For example, without additional preprocessing, the key `CPUs` will
+       become `cp-us` if put through a caramelize with being preformatted to `Cpus`.
+    """
+    if "CPUs" in key:
+        key = key.replace("CPUs", "Cpus")
+    key = _acronym.sub(r"_", key)
+    return _camelize.sub(r"_", key).lower()
+
+
+def generate_descriptors(opt: str):
+    """Generate descriptors for retrieving and mutating configuration options.
+
+    Args:
+        opt: Configuration option to generate descriptors for.
+    """
+
+    def getter(self):
+        return self.data.get(opt, None)
+
+    def setter(self, value):
+        self.data[opt] = value
+
+    def deleter(self):
+        del self.data[opt]
+
+    return getter, setter, deleter
+
+
+class LineInterface:
+    """Interface for data models that can be constructed from a configuration line."""
+
+    @classmethod
+    @abstractmethod
+    def from_str(cls, line: str):
+        """Construct data model from configuration line."""
+
+    @abstractmethod
+    def __str__(self) -> str:
+        """Return model as configuration line."""
+
+
+class BaseModel(ABC):
+    """Base model for Slurm data models."""
+
+    def __init__(self, validator=None, /, **kwargs) -> None:
+        for k, v in kwargs.items():
+            if not hasattr(validator, k):
+                raise ModelError(
+                    (
+                        f"unrecognized argument {k}. "
+                        + f"valid arguments are {[opt.name for opt in validator]}"
+                    )
+                )
+
+        self.data = kwargs
+
+    @classmethod
+    def from_dict(cls, data: Dict[str, Any]):
+        """Construct new model from dictionary."""
+        return cls(**data)
+
+    @classmethod
+    def from_json(cls, obj: str):
+        """Construct new model from JSON object."""
+        data = json.loads(obj)
+        return cls.from_dict(data)
```

Comment:
Nit: this can be typed with a typevar:
```Python
TBaseModel = TypeVar("TBaseModel", bound="BaseModel")

def from_dict(cls: Type[TBaseModel], data: Dict[str, Any]) -> TBaseModel
```
But we can also skip this if you think it's getting too weird for just type annotations 😅

Reply:
I'd consider this as out of scope for this PR. 

`slurmutils` overall needs proper type checking added to it, so I'm expecting that this won't be the only place where we need to add better and more accurate type annotations :sweat_smile: